src/robot/client.ts

```diff
@@ -376,6 +381,7 @@ export class RobotClient extends EventDispatcher implements Robot {
   }
 
   public async disconnect() {
+    this.emit(MachineConnectionEvent.DISCONNECTING, {});
```

Comment:
(q) is this necessary, given that we've set `this.on(eventType ...)` above?

Same with other `this.emit` calls below.

Reply:
Similar (q): why not just emit the `connectionstatechange` here, rather than emitting this event, then catching it and re-emitting?

Reply:
This is actually emitting the source event, signaling disconnect is starting. The `this.on(eventType, ...)` above re-emits the event with the name `connectionstatechange`. What this means is the consumer can choose which event they want to listen to - if they care about all connection events, they can set up a `robotClient.on('connectionstatechange', ...)` event handler. If they only want to listen to disconnected events, they can just set up a `robotClient.on('disconnected', ...)` handler.

We don't want to just emit `connectionstatechange` here because we would need to emit the `connectionstatechange` anywhere a disconnecting, disconnect, connecting, or connected event is emitted. By catching and re-emitting, we don't need to remember to also emit `connectionstatechange` everywhere.